mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala

```diff
@@ -512,6 +513,92 @@ object Vectors {
     squaredDistance
   }
 
+  private def dot(a : DenseVector, b : DenseVector) : Double = {
+    (a.toArray zip b.toArray).map(x => (x._1 * x._2)).sum
+  }
+
+  private def dot(a : SparseVector, b : DenseVector) : Double = {
+    (a.indices zip a.values).map(x => { b(x._1)*x._2 }).sum
```

Comment:
Ah... I keep forgetting that part, of course. I think that `Vector` is really an API for Spark itself, and Spark doesn't have this problem to solve as a result since it can access Breeze, etc. However I can see that `Vector` is something user code reasonably uses and manipulates. While it may be best for apps to put these into their own desired representation for any serious manipulation, there's an argument that giving more than the bare essentials in an API is worth the effort. I suppose that's the question is SPARK-6442. dot-product seems like a legitimate question; Hadamard, not so sure.